ENH: add zonal_stats

masawdah opened the pull request:
Thanks for providing this valuable package. The idea of this PR to aggregate raster data cube over a set of polygon geometries  #35 as it needed to use within openEO process Open-EO/openeo-processes-dask#124  . It is based on dask array and parallelization which make it efficient and fast. It has been tested to aggregate polygons over data on continental level and shows a good performance.  

I'd like to know your thoughts about that. 

Here is an example: 
```
import datetime
import geopandas as gpd
import pandas as pd
import xarray as xr
import numpy as np
import xvec
from shapely.geometry import Polygon
```
```
# Create a dataset with 2 variables and 3 time steps 
np.random.seed(0)

temperature = 15 + 8 * np.random.randn(20, 20, 3)
precipitation = 15 + 10 * np.random.randn(20, 20,3)
lat = np.linspace(30,40,20)
lon = np.linspace(10,20,20)



time = pd.date_range("2014-09-06", periods=3)
reference_time = pd.Timestamp("2014-09-05")


ds = xr.Dataset(
    data_vars=dict(
        temperature=(["x", "y", "time"], temperature),
        precipitation=(["x", "y", "time"], precipitation),
    ),
    coords=dict(
        x=lon,
        y=lat,
        time=time,
        reference_time=reference_time,
    ),
    attrs=dict(description="Weather related data."),
)
```

```
ds = ds.chunk(dict(x=4,y=4))
```


```
# Create geometries over the dataset
num_polygons = 2  # Adjust the number of polygons as needed
polygons = []

for _ in range(num_polygons):
    # Generate random polygon coordinates within the bounding box of the downsampled dataset
    lon = np.random.uniform(ds.x.min(), ds.x.max(), 4)
    lat = np.random.uniform(ds.y.min(), ds.y.max(), 4)
    polygons.append(Polygon(zip(lon, lat)))


geoseries = gpd.GeoSeries(polygons)
gdf = gpd.GeoDataFrame(geometry=geoseries)

gdf = gdf.set_geometry('geometry')
gdf.crs = 'EPSG:4326'

polys = gdf.geometry.values
```
```
extracted = ds.xvec.agg_polys(polys, stat="mean")
```
![image](https://github.com/xarray-contrib/xvec/assets/61426508/fd4b5d85-a2ec-429b-bf1a-3957cf82811e)

martinfleis commented:
Thanks! I'll have a look later but note that you need to add new deps to https://github.com/xarray-contrib/xvec/tree/main/ci

martinfleis commented on xvec/accessor.py:
```diff
@@ -918,6 +929,128 @@ def to_geodataframe(
         )
         return df
 
+    def agg_geom(self, geom, trans,var,  stat='mean'):
```
Could you add a short docstring?

masawdah replied:
Added.

martinfleis commented on xvec/accessor.py:
```diff
@@ -918,6 +929,128 @@ def to_geodataframe(
+
+    def spatial_agg(self, geometries, stat='mean', chunk_size = 2):
+        transform = self._obj.rio.transform()
+        num_cores = multiprocessing.cpu_count() 
```
Shall we expose this as `n_jobs` defaulting to `-1` (all threads)?

masawdah replied:
I made it an optional user input with -1 as default.

martinfleis commented on xvec/accessor.py:
```diff
@@ -918,6 +929,128 @@ def to_geodataframe(
+    
+
+
+    def spatial_agg(self, geometries, stat='mean', chunk_size = 2):
```
small docstring?

masawdah replied:
Added.

martinfleis commented on xvec/accessor.py:
```diff
@@ -918,6 +929,128 @@ def to_geodataframe(
+        return vec_cube
+
+    
+    def agg_polys(
```
we should pick a better name. what is `{stars}` using?

edzer replied:
In R, `aggregate` is a generic, `{stars}` has a [method for stars objects](https://r-spatial.github.io/stars/reference/aggregate.stars.html)

martinfleis replied:
Both `rasterstats` and `xarray_spatial` use `zonal_stats,` so it may be best to use the same?

masawdah replied:
Renamed to zonal_stats.

martinfleis reviewed (commented):
Hey, a first pass, so far without running the code. A few notes below and some others in the code.

Do we need `dask` here? I mean, optimally we make use of it if available but don't require it for smaller use cases.

If possible, I would like to avoid storing data in the repository here. We can use `geodatasets` to fetch us counties and xarray example data to get some raster cubes.

Can you ensure that pre-commit checks pass?

masawdah commented:
I'll go through the notes and work on them.

masawdah commented on ci/310.yaml:
```diff
@@ -6,6 +6,11 @@ dependencies:
   - xarray
+  - rioxarray=0.14.1
+  - joblib=1.3.2
+  - rasterio=1.3.8
```
Done

masawdah commented on ci/dev.yaml:
```diff
@@ -7,6 +7,11 @@ dependencies:
   - proj
+  - rioxarray=0.14.1
+  - joblib=1.3.2
+  - rasterio=1.3.8
```
Done

masawdah commented on ci/latest.yaml:
```diff
@@ -6,6 +6,11 @@ dependencies:
   - xarray
+  - rioxarray=0.14.1
+  - joblib=1.3.2
+  - rasterio=1.3.8
```
Done

masawdah commented on environment.yml:
```diff
@@ -6,6 +6,9 @@ dependencies:
   - xarray
+  - rioxarray=0.14.1
+  - joblib=1.3.2
+  - rasterio=1.3.8
```
Done

masawdah commented on xvec/accessor.py:
```diff
@@ -9,9 +9,20 @@
+import rasterio
+from geopandas import GeoDataFrame
+import geopandas as gpd
+import gc
```
Done. Import them just we they're needed.

masawdah commented:
Hey, I went through the notes. Now the code passed the pre-commit. And I made it an optional to use dask and number of jobs if the data is big as the following: 

 ```
extracted = ds.xvec.zonal_stats(polys, stat="mean", dask = False, n_jobs = -1)
```

 ```
ds1 = ds.chunk(dict(x=4,y=4))
extracted = ds1.xvec.zonal_stats(polys, stat="mean", dask = True, n_jobs = -1)
```

And I think it's possible to use geodatasets  and xarray example data to get some raster cubes instead of storing the data in the repo. I'll look into that and let u know.

masawdah commented:
I've tested it using counties fetched from geodatasets and  'air_temperature' data cube. It works but the order of (x, y ) dimensions is not consistent with what implemented now (similar to data cube structure in openEO) so I had to modify the order manually. 

We can add function to make sure the dimension order is as expected. What do u think ? 

```
from geodatasets import get_path
counties = gpd.read_file(get_path("geoda.natregimes"))
polys = counties.geometry.values

ds = xr.tutorial.open_dataset("air_temperature")

extracted = ds.xvec.zonal_stats(polys[:2], stat="sum", dask = False, n_jobs = -1)

extracted
```
![image](https://github.com/xarray-contrib/xvec/assets/61426508/4831802c-54c9-4f34-9b99-aa9d7c1e2163)

martinfleis commented:
> We can add function to make sure the dimension order is as expected. What do u think ?

In `extract_points`, we have the dimension mapping exposed directly as `x_coords` and `y_coords` (https://xvec.readthedocs.io/en/stable/generated/xarray.Dataset.xvec.extract_points.html). We could maybe use something like that?

masawdah commented:
> > We can add function to make sure the dimension order is as expected. What do u think ?
> 
> In `extract_points`, we have the dimension mapping exposed directly as `x_coords` and `y_coords` (https://xvec.readthedocs.io/en/stable/generated/xarray.Dataset.xvec.extract_points.html). We could maybe use something like that?

I'll take a look at that. But for now I made (lon, lat) as default dims names to aggregate over them. And added a function to rename any other possible names (in a dictionary) to (lon, lat) to be consistent with the function requirement.  So, now it works for both data cubes from openEO and data cubes from xarray.

masawdah commented:
> > We can add function to make sure the dimension order is as expected. What do u think ?
> 
> In `extract_points`, we have the dimension mapping exposed directly as `x_coords` and `y_coords` (https://xvec.readthedocs.io/en/stable/generated/xarray.Dataset.xvec.extract_points.html). We could maybe use something like that?

I think  it is saver to do it in this way, the name of the coordinates or its order "as dask does not support the name "  https://docs.dask.org/en/stable/generated/dask.array.sum.html

masawdah commented:
I've added the dimension mapping like what used for "extract_points", with option to handle dask dims. 

 ```
from geodatasets import get_path
import geopandas as gpd
import xarray as xr
import xvec
import dask

counties = gpd.read_file(get_path("geoda.natregimes"))
ds = xr.tutorial.open_dataset("air_temperature")

extracted = ds.xvec.zonal_stats(polys[:2], x_coords='lon',y_coords='lat', stat="mean", n_jobs = -1)
 ```
![image](https://github.com/xarray-contrib/xvec/assets/61426508/27b6dc6f-2459-4e29-9849-ea946646588f)

 ```
ds2 = ds.chunk(dict(lon=4,lat=4))
extracted = ds.xvec.zonal_stats(polys[:2], x_axis=2, y_axis=1, stat="sum", dask = True, n_jobs = -1)
 ```
![image](https://github.com/xarray-contrib/xvec/assets/61426508/6233e28d-d18a-412c-9a34-a32925b135be)

martinfleis commented on xvec/accessor.py:
```diff
@@ -918,6 +918,375 @@ def to_geodataframe(
         )
         return df
 
+    def agg_geom(
```
```suggestion
    def _agg_geom(
```
This should be private, the public API is `zonal_stats`.

martinfleis commented on xvec/accessor.py:
```diff
@@ -918,6 +918,375 @@ def to_geodataframe(
+            Aggregated values over the geometry.
+
+        """
+        try:
+            import geopandas as gpd
+        except ImportError as err:
+            raise ImportError(
+                "The geopandas package is required for `xvec.agg_geom()`. "
+                "You can install it using 'conda install -c conda-forge geopandas' or "
+                "'pip install geopandas'."
+            ) from err
```
We don't need GeoPandas here. We just need some array of shapely geometries.

martinfleis commented on xvec/accessor.py:
```diff
@@ -918,6 +918,375 @@ def to_geodataframe(
+                "'pip install rasterio'."
+            ) from err
+
+        try:
+            import gc
+        except ImportError as err:
+            raise ImportError(
+                "The gc package is required for `xvec.agg_geom()`. "
+                "Make sure 'gc' is included in the standard library"
+                "Check your Python installation for any issues."
+            ) from err
```
```suggestion
        import gc
```
No need to check for built-in module.

martinfleis commented on xvec/accessor.py:
```diff
@@ -918,6 +918,375 @@ def to_geodataframe(
+                "Check your Python installation for any issues."
+            ) from err
+
+        # Create a GeoSeries from the geometry
+        geo_series = gpd.GeoSeries(geom)
+
+        # Convert the GeoSeries to a GeometryArray
+        geometry_array = geo_series.geometry.array
```
```suggestion
        geometry_array = np.asarray(geom)
```

As mentioned above, we don't need to use geopandas.

masawdah replied:
I've recheck it and found that this step not important and removed it completely.

martinfleis commented on xvec/accessor.py:
```diff
@@ -918,6 +918,375 @@ def to_geodataframe(
+        )
+
+        masked_data = xar_chunk * mask[:, :, np.newaxis]
+        del mask, xar_chunk
+        gc.collect()
```
How important this is in your opinion?

masawdah replied:
It's important in case of limited resources. I've tested it over continental level "sentinel-2 tiles" for 150 polygons and the code crashed without it. It helps to save memory when have big data.

martinfleis replied:
Okay!

martinfleis commented on xvec/accessor.py:
```diff
@@ -918,6 +918,375 @@ def to_geodataframe(
+
+        return result
+
+    def spatial_agg(
```
```suggestion
    def _spatial_agg(
```
This also should be private.

martinfleis commented on xvec/accessor.py:
```diff
@@ -918,6 +918,375 @@ def to_geodataframe(
+                "'pip install tqdm'."
+            ) from err
+
+        try:
+            import gc
+        except ImportError as err:
+            raise ImportError(
+                "The gc package is required for `xvec.spatial_agg()`. "
+                "Make sure 'gc' is included in the standard library"
+                "Check your Python installation for any issues."
+            ) from err
```
```suggestion
        import gc
```

martinfleis reviewed (changes requested):
Sorry for a delay! Notes below.


If I understand it correctly, we are parallelising the code twice. Once using `Parallel` and then, optionally using Dask. The latter is done by calling `dask.array`. I am wondering if this should be handled completely differently. 

If we have an xarray Dataset that is chunked using Dask, then when calling `sum` or `mean`, we get parallelisation for free, no?  So we don't need to interact with Dask ourselves. (@benbovy is that correct?)

If we don't have a chunked array, then we may want to do something ourselves. In that case, I believe it makes sense to use `joblib`. Assuming that most polygons in reality fall within a single chunk of a chunked array, it is probably okay to use it also when we have a chunked array. Though I can imagine a situation when geometries often cross multiple chunks, leading to `joblib` trying to use all available cores while `Dask` is trying to do the same.

My suggestion is to completely remove all of `if dask:` cases. The benefit of using dask in this way is imho limited and if your data is chunked, you get dask aggregation anyway. That will also simplify the signature.

---

Please remove the `data` folder. It is not used anywhere and I'd like to avoid storing any data in the repository. If you need to real-world data to be used, I suggest using xarray datasets for raster and geodatasets for vector data or creating dummy data on-the-fly.

---

Apologies again for a late review but I didn't want to rush this. I also see some follow-ups I'd like to have in xvec but let's leave those out of this PR.

If you can implement the suggestions above we can then iterate over the functionality in issues and PRs to make it as great as it can be :).

Thanks! And sorry for a lot of notes :).

masawdah commented:
Thanks for your review and feedback. I reviewed the suggestions and made the necessary modifications to the code accordingly.

martinfleis commented on xvec/accessor.py:
```diff
@@ -918,6 +918,274 @@ def to_geodataframe(
+        for k in stats_dic.keys():
+            s = stats_dic[k]
+            columns = []
+            for t in range(len(self._obj.time)):
```
Why do we have `time` hard-coded here? This should be agnostic to other dimensions. The same hard-coding is also below.

masawdah replied:
That's correct. I did it based on the example of raster cube. I'll make the necessary adjustments to ensure it's dimension-agnostic.

martinfleis commented on xvec/accessor.py:
```diff
@@ -918,6 +918,274 @@ def to_geodataframe(
+        x_coords: Hashable,
+        y_coords: Hashable,
+        stat: str = "mean",
+        name: str = "geometry",
```
This is not propagated to `_spatial_agg`

masawdah commented:
Hi , I've adjusted the code to be more flexible and dimension-agnostic, as you can see below. I've just hard-coded the name of one coordinate as 'data_variables' to save the data variable names temporary, but we can make it a user input. However, I believe this is not critical, as it's an internal step that doesn't affect the output, unless we intend to have the output as an xarray.DataArray.



![image](https://github.com/xarray-contrib/xvec/assets/61426508/47874a51-07b2-427f-ba39-bb9afb1a0c3c)

![image](https://github.com/xarray-contrib/xvec/assets/61426508/e61d8334-7f8f-4578-ae1c-bb06f0faa90c)

martinfleis commented:
Hey, I did a deep dive into this which resulted in refactoring of the code. I opted to use more xarray logic on the process than before, which simplifies the code quite a bit. 

I also noticed that the implementation was providing wrong results when I compared it to the one based on `geocube` so I also fixed that and since I have a geocube-based Implementation now I will make a follow-up PR exposing it as another engine. It is faster but not suitable for small polygons. Hopefully, we can also expose `exactextract` once the Python bindings are released (a prototype is there).

I have removed chunking as I believe that is something a user can do if they run into issues but it was more of a bottleneck in most use cases I tried.

Plus some other changes you can see in the code.

I plan on adding tests etc and then merge. I will follow-up with the geocube version, user guide notebook and a release.

martinfleis commented:
Looking at it in more detail, we can probably use `rasterio.features.rasterize` ourselves without depending on `geocube`. I'll look into that during the follow up.

martinfleis reviewed (approved):
Going to merge this now and will continue in separate PRs with the remaining stuff.

@masawdah thanks for starting this, my refactor was way easier than if I'd be starting from scratch! Let me know if there are some specific requirements on the openEO side you'd like to see included before a release.

masawdah commented:
Sounds great,  looking forward to the new release :)
